### PR TITLE
build: setup contrib

### PR DESCRIPTION
**Pull request by mattklein123**
Also move Squash extension to contrib

As per: https://docs.google.com/document/d/1yl7GOZK1TDm_7vxQvt8UQEAu07UQFru1uEKXM6ZZg_g/edit#

Part of https://github.com/envoyproxy/envoy/issues/14078

Risk Level: Low
Testing: Existing tests. Also a bunch of manual testing.
Docs Changes: Added
Release Notes: Added
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
@phlax @lizan I have more work to do on documentation but I think this is pretty close to ready on the code side. Can you give me a first pass to see if you have any high level comments while I work on  more docs?

**Comment by mattklein123** [comment]
@phlax @htuch @adisuissa @lizan this is ready for review. PTAL. The only thing not done is the creation of the docker hub repos which I will do once we agree on the general shape of this.

File: contrib/filters/http/squash/source/config.cc
Comment (lizan) [inline contrib/filters/http/squash/source/config.cc:1]: I thought this will be `contrib/squash/source/filters/http/config.h`? Per https://docs.google.com/document/d/1yl7GOZK1TDm_7vxQvt8UQEAu07UQFru1uEKXM6ZZg_g/edit#heading=h.zgixgx1c4t8u

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: IMO it makes more sense to use the full extension path that matches the normal extensions tree. I can make the doc more clear on this.

**Review by adisuissa — COMMENTED** [review]
Changes LGTM, thanks!
Left a couple of suggestions here and in the doc.

File: contrib/filters/http/squash/source/config.cc
Comment (lizan) [reply contrib/filters/http/squash/source/config.cc:1]: I meant we want contrib/<extension_name>/ as a unit, have source and test under it. In that way we can group extensions that have to be used together (for example, something like dynamic_forward_proxy should have both filter and cluster extension under `contrib/dynamic_forward_proxy`.) It also makes CODEOWNER clearer instead of mixing multiple extensions in same tree.

File: EXTENSION_POLICY.md
Comment (htuch) [inline EXTENSION_POLICY.md:146]: Please emphasize lack of security coverage here, SECURITY.md and also https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model#core-and-extensions.

It really should be caveat emptor.

File: tools/proto_format/proto_sync.py
Comment (htuch) [inline tools/proto_format/proto_sync.py:167]: I'd prefer tooling that enforces that everything in `contrib/` is by default `v3alpha`. Then we can explicitly allowlist the few exceptions (grandfathered or agreed on API shepherd coverage). This will prevent accidental leakage of contrib APIs into stable.

File: contrib/filters/http/squash/source/config.cc
Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: I did end up grouping source/test under the same name, but I still think if we just do things by name it will get pretty chaotic. I have a strong preference for matching the same directory structure as source/extensions. You don't think it will get extremely messy if we don't have this structure?

**Comment by mattklein123** [comment]
@htuch @lizan @adisuissa updated per comments

File: contrib/filters/http/squash/source/config.cc
Comment (lizan) [reply contrib/filters/http/squash/source/config.cc:1]: > I did end up grouping source/test under the same name, but I still think if we just do things by name it will get pretty chaotic. I have a strong preference for matching the same directory structure as source/extensions. You don't think it will get extremely messy if we don't have this structure?

No, that was the whole point I proposed `contrib/<extension_name>` and I feel strongly that under every `contrib/extension_name` it should look like a standalone repository. Then we can give the CODEOWNERS responsibility of managing everything under it, if no one is maintaining we can just delete the whole directory.  It also has the flexibility of having multiple extension under same name like I stated above. We can have a guidance that under `contrib/<extension_name>` should still follow same structure of the main tree.

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: I think this falls apart for when contrib extensions share common code, which they will immediately I think when moving both postgres and myself. So then what is done with common code?

I'm not going to die on this hill as I would like to get this merged, but I don't think your proposal will work in all cases anyway, so I'm not sure it's worth changing to it.

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: (Also, API is in another directory anyway.)

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: (And also I just think it's confusing to have a top level directory of names with no organization. I think it will become unwieldy quickly.)

cc @envoyproxy/maintainers for further thoughts.

Comment (ggreenway) [reply contrib/filters/http/squash/source/config.cc:1]: With regards to API files, I think they should also be co-located with the source and tests for the extension, not in a separate `api` tree.

As far as structure, I think I agree with @lizan. Given the governing/social contract we're making for `contrib`, I think we should have a bunch of completely independent silo directories. For shared code, I think we could have another silo of `db_common` or something for shared postgres/mysql code (or similar); it's owners could be the union of the codeowners of each of those two extensions. Or each extension could have it's own entire copy of the shared code; that may end up being my preference.

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: > With regards to API files, I think they should also be co-located with the source and tests for the extension, not in a separate api tree.

I tried to do this but it's not possible without a tremendous amount of work due to all of the API machinery. I'm sure it's possible but it's not work I have time to do, so if we want that someone else will need to pick it up. Note that I already cleared this with the API shepherds.

> Or each extension could have it's own entire copy of the shared code; that may end up being my preference.

This is a hill I'm willing to die on. I don't think code copying makes sense, so we must have some way of sharing code. For me personally I really don't understand how a giant top level mashed you tree is preferable to a nicely organized tree. Owners files can be used in either case, and we can track status in either case. So why not have nice organization?

As I said, I'm not going to demand the existing structure, though I strongly recommend it. I am going to demand some way of sharing code, so let me know what you all prefer there.

Comment (htuch) [reply contrib/filters/http/squash/source/config.cc:1]: I think the API stuff could be made to work colocated, but it's no big deal if we have it as independent. 

If structure is determined by extension structure, and there aren't any new extension types in `contrib/`, I think I'd be inclined to vote to mirror existing `source/extensions` structure. I think we should give folks in `contrib/` maximum freedom, but not where it can only lead to unnecessary disorganization.

File: EXTENSION_POLICY.md
Comment (htuch) [inline EXTENSION_POLICY.md:144]: Tiny nit: s/"contrib."/`contrib/`./

File: CONTRIBUTING.md
Comment (htuch) [inline CONTRIBUTING.md:273]: Where is the `v3alpha` policy (or API shepherd coverage) articulated?

File: contrib/filters/http/squash/source/config.cc
Comment (lizan) [reply contrib/filters/http/squash/source/config.cc:1]: > As I said, I'm not going to demand the existing structure, though I strongly recommend it. I am going to demand some way of sharing code, so let me know what you all prefer there.

Regarding to sharing code, any sharing between contrib and core goes to core, so the question is how to share codes between contrib extensions. I'd vote to have another silos in `contrib/common/sqlutils`, this is no different from `source/extensions/common`. Looking at existing [`source/extensions/common`](https://github.com/envoyproxy/envoy/tree/main/source/extensions/common) probably only `sqlutils` will go contrib/common like that. For example even if we move redis to contrib (while we're not going to do), in the model that I proposed, the common code, network filter, health checker and cluster will be all under `contrib/redis`. The tree under `contrib/redis` can still match the existing structure though.

File: contrib/contrib_build_config.bzl
Comment (ipuustin) [inline contrib/contrib_build_config.bzl:7]: Just so that I understand this right: `bazel --//contrib/filters/http/squash:enabled ...` would enable this contrib extension from the bazel command line when building something else than `//contrib/exe/envoy-static` build target?

File: contrib/filters/http/squash/source/config.cc
Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: > I'd vote to have another silos in contrib/common/sqlutils, this is no different from source/extensions/common

Yes, I agree, this is what I was going to do when I move that code.

> I proposed, the common code, network filter, health checker and cluster will be all under contrib/redis. The tree under contrib/redis can still match the existing structure though.

I understand the proposal, I just don't agree that it's worth making that directory incredibly messy. It's an extreme counterpoint, but imagine if all of the linux kernel drivers were in a single directory sorted by name of the driver only? That's what is being proposed here. I will try to ping all the maintainers in Slack so we can close this out. If the group consensus is to use this structure so be it.

File: contrib/contrib_build_config.bzl
Comment (mattklein123) [reply contrib/contrib_build_config.bzl:7]: I'm not sure. I need to test this. I'm pretty sure it won't work outside of this target, but it will work when building the contrib exe to customize that build. I will try it out and report back.

File: CONTRIBUTING.md
Comment (mattklein123) [reply CONTRIBUTING.md:273]: It's in the linked google doc, but I will mention it here also.

File: contrib/filters/http/squash/source/config.cc
Comment (phlax) [reply contrib/filters/http/squash/source/config.cc:1]: i think i tend to agree with @lizan altho separating every extension into its own directory means a lot of directory duplication in the envoy source, i think it probs makes more sense from the pov of extension builders to see examples that are entirely independent

i guess i get the point about linux kernel modules, but at the same time i think that is more of a comparison to core extensions rather than contrib ones

i guess there a couple of different motivations here - one is the maintainability/sanity of the envoy contrib source, and another is isolating extensions to make it clearer how the modularisation works to devs - if the latter is better served one way or other im inclined towards that way

if we do go down the path of separating each extension, there is also a question of extensions that extend other extensions and whether they then can/should be grouped

Comment (alyssawilk) [reply contrib/filters/http/squash/source/config.cc:1]: I mildly prefer keeping code for a given extension in a given directory, though deleting the contrib/source/extensions/foo contrib/include/extensions/foo contrib/test/extensions/foo isn't a deal breaker (I'm fine either way)

I'm strongly against copy-pasted common code.  I'd lean towards a common directory with default visibility private, and explicit opt-in for each library so we can easily tell that common/foo_lib is used by extension A and B, and clean up if we delete extension A and B.

Comment (mattklein123) [reply contrib/filters/http/squash/source/config.cc:1]: Sorry I want to be completely clear about what is being proposed here since I'm not sure everyone is talking about the same thing.

The current code:

`/contrib/<... full hierarchical path based on extension type ...>/<extension name>/<source|test>/`

Lizan's proposal:

`/contrib/<extension name>/<... full hierarchical path based on extension type ...>//<source|test>/`

The latter means that when one looks in the contrib folder they will see a giant list of names with no context beyond that.

File: contrib/contrib_build_config.bzl
Comment (mattklein123) [reply contrib/contrib_build_config.bzl:7]: `bazel build //contrib/exe:envoy-static --//contrib/filters/http/squash/source:enabled=false` works properly, but trying to enable the contrib filter this way on top of `//source/exe:envoy-static` does not work. I will make this clear in the docs.

File: contrib/filters/http/squash/source/config.cc
Comment (ggreenway) [reply contrib/filters/http/squash/source/config.cc:1]: The reason I suggested copy/paste shared code, is that if it isn't clear who the owner of the shared code is, what happens if one of the extensions using it wants to make a change and the other user doesn't? But I'm fine with the group-consensus to not duplicate the shared code.

Comment (ggreenway) [reply contrib/filters/http/squash/source/config.cc:1]: I think one way to looking at the `/contrib/extension-name/full` is like github organizations. The `extension-name` part is like a github org; then there's >=1 extensions underneath it, all owned by the same logical owner.

I think both proposals have merit; I'd be fine with either approach, and as we get more experience with managing this, and if we feel like it, we can re-org it into the other approach if needed. So do whatever is most expedient.

**Comment by mattklein123** [comment]
Updated, PTAL

**Review by ggreenway — APPROVED** [review]
Docs, policy, directory structure all look good to me. Will defer to other reviewers on the bazel/build-system changes.

**Comment by mattklein123** [comment]
@lizan @htuch PTAL when you have time, thanks.

**Review by lizan — APPROVED** [review]

**Review by htuch — COMMENTED** [review]
Post-merge LGTM, thanks!

**Pull request merged**
